Initial overview

tlively opened the pull request.

binji commented on proposals/feature-detection/Overview.md:
```diff
@@ -2,8 +2,141 @@
+   respective features.
+ - Modules should be able to support multiple overlapping feature sets with
+   minimal code duplication or bloat.
+ - It is an explicit non-goal to allow multiversioned modules to validate on
```
Yeah, I think this is OK. It'd be nice to fall back to MVP, but I think if we encourage support of this feature we can at least reduce the number of wasm modules from N to 2 (and eventually 1).

binji commented on proposals/feature-detection/Overview.md:
```diff
@@ -2,8 +2,141 @@
+introduced. These sections contain extensions to previous sections such as the
+type section or code section, but their contents are only considered to be part
+of the module during validation and instantiation if the *active feature set
+index* is in the sections' set of activating feature sets. Feature-conditional
```
I read this meaning you can only have one feature-conditional section, and the known section must exist. Is that the intention? Or can you have the following:

```
type
conditional-type
conditional-type
```

and

```
(no type section)
conditional-type
```

tlively replied:
Good question. I suppose the latter should be ok as well, but then we would have to put the section ID of the extended section in each of the feature-conditional sections as well. It might allow smaller code size if we simply require a zero-length unconditional section in that case.

tlively replied:
@binji Do you think it would make more sense to specify the conditional sections as subsections of the sections they extend instead? Engines would be able to detect their presence by checking whether there is additional space left in the section after the MVP contents.

binji replied:
Interesting... that could work, I think. The only issue I can see is that you wouldn't be able to conditionally include a new section.

tlively replied:
I think you could still just have a zero length vector for the unconditional contents. I think this formulation is clearer, so I’ll rewrite the description for this.

binji replied:
I mean, what if we introduce a new section id with a feature proposal. Then an older engine would fail to load that binary, even though the section was conditional.

binji commented on proposals/feature-detection/Overview.md:
```diff
@@ -2,8 +2,141 @@
+ - Import section
+ - Export section
+ - Start section
+ - Element section
```
Not sure I buy this -- we already have extended the element and data sections for the bulk-memory proposal, we may want to modify them again.

tlively replied:
I was thinking that any changes to these sections (e.g. to use passive segments) would require a change in the host interface, but if you do all your memory.inits in the start function, that's not true. On the other hand, there's no reason to use passive segments over active segments that doesn't require a change to the module interface, such as importing a shared memory instead of an unshared memory.

binji replied:
I'm hesitant to guess why we'd want to make changes to these sections. For example, we're already planning to use the element section to declare references in the reference-types proposal.

binji commented on proposals/feature-detection/Overview.md:
```diff
@@ -2,8 +2,141 @@
+ change the module interface, would not be meaningful, or would be complex and
+ have no obvious benefit:
+
+ - Import section
```
We're already going to extend import/export section for the exception-handling proposal, this would make it more difficult to provide a fallback.

tlively replied:
I'm not sure it makes sense to provide a fallback for a case like that. The presence of exceptions is essentially a breaking change in the ABI of the module and would necessitate pervasive code changes and therefore extensive duplication, and also a difference in the interface with the host runtime.

binji replied:
Exceptions are just one example. The [type imports proposal](https://github.com/WebAssembly/proposal-type-imports) is one where it may make sense to have a fallback for engines that don't support type exports.

binji commented on proposals/feature-detection/Overview.md:
```diff
@@ -2,8 +2,141 @@
+
+| Field        | Type           |
+|--------------|----------------|
+| feature sets | `feature_set*` |
```
It seems like this could lead to combinatorial explosion. Though I suppose in reality the tools will never produce 2**n different modules.

tlively replied:
I expect users will have to explicitly opt in to each feature set that is produced. Tools should not eagerly adding new feature sets without explicit buy-in from the user.

tlively commented on proposals/feature-detection/Overview.md:
```diff
@@ -2,8 +2,141 @@
+features in a backward-compatible manner, i.e. without running afoul of older
+validation rules.
+
+The goal of this proposal is to allow functions to be implemented differently
```
Yes, and maybe globals as well, but I'm not convinced of that.

rossberg replied:
Well, what you describe below already allows it, so I thought this sentence simply is out of sync. (FWIW, a use case would be having a different type for a global if certain features are present, especially ones introducing new types.)

tlively commented on proposals/feature-detection/Overview.md:
```diff
@@ -2,8 +2,141 @@
+ - Start section
+ - Element section
+ - Data count section
+ - Data section
```
I have not been able to think of a use for extending these sections that does not also require changing the module interface. Perhaps we should allow extensions to these sections as well so users are not constrained by my lack of imagination :)

tlively commented:
> IIUC, with the design described here, feature sets are defined globally upfront, and for each section that depends on the chosen feature set, N conditional sections need to be provided -- even if only two of them differ for the respective kind of section. That would potentially lead to substantial code duplication.

The *activating feature sets* vector in each conditional section lets it be activated by multiple conditional sections, so there's no need to emit identical conditional sections.

> I think it makes more sense to put the feature sets in each of the conditional sections themselves. That way, each section kind can distinguish feature sets at a granularity independent from other sorts of sections. That also is more akin to how conditional compilation is used in, say, C.

My intention was to allow this granularity, but require the feature sets to be declared up front to avoid repeating all those strings everywhere.

> That is, instead of a feature section we simply have conditional compilation, and the condition consists of a list of strings naming present features. This scheme could potentially be extended by allowing to "set" custom strings at instantiation time (or even unset predefined ones), leading to a more versatile conditional compilation mechanism.

I think the current scheme would also support custom strings as an extension in the JS API, but unless we have people asking for it I think we should consider that to be out of scope for now.

rossberg commented:
> The activating feature sets vector in each conditional section lets it be activated by multiple conditional sections, so there's no need to emit identical conditional sections.

Ah, I see, thanks.

>> I think it makes more sense to put the feature sets in each of the conditional sections themselves. That way, each section kind can distinguish feature sets at a granularity independent from other sorts of sections. That also is more akin to how conditional compilation is used in, say, C.
> My intention was to allow this granularity, but require the feature sets to be declared up front to avoid repeating all those strings everywhere.

Isn't that premature micro optimisation? How many strings do we expect, and what is their size relative to the code they are guarding (or the size of the module in general)?

It would obviously be simpler without the extra section type and no extra indirection to go through.

Another question: Right now the binary format has the useful property that you can merge modules by simply concatenating sections plus renumbering index spaces. Is that property maintained under the proposed design? (I think it may be, but I might be missing some subtlety.)

tlively commented:
> > > I think it makes more sense to put the feature sets in each of the conditional sections themselves. That way, each section kind can distinguish feature sets at a granularity independent from other sorts of sections. That also is more akin to how conditional compilation is used in, say, C.
> > > My intention was to allow this granularity, but require the feature sets to be declared up front to avoid repeating all those strings everywhere.
> 
> Isn't that premature micro optimisation? How many strings do we expect, and what is their size relative to the code they are guarding (or the size of the module in general)?

Possibly. Say a module had the following feature sets:

 - "nontrapping-fptoint", "bulk-memory", and "tail-call"
 - "nontrapping-fptoint" and "bulk-memory"
 - "nontrapping-fptoint"

That's just under a hundred bytes times however many times they have to show up in the file, which is at least once for the function section and once for the code section. Whether that's worth optimizing is a matter of taste.

Another nice property of the separate feature set section is that the active feature set is decided on once and consistently applied for all conditional sections. Without that mechanism we would have to take the first applicable conditional chunk for each section type and we would not have a way to include multiple conditional chunks for a given section. I don't think this is a very important property, though.

> It would obviously be simpler without the extra section type and no extra indirection to go through.

Agreed. Lets go with the simpler solution for now until we have data or concrete use cases justifying the extra indirection.

> Another question: Right now the binary format has the useful property that you can merge modules by simply concatenating sections plus renumbering index spaces. Is that property maintained under the proposed design? (I think it may be, but I might be missing some subtlety.)

No, this property is not maintained in general. Consider `a.wasm` with an unconditional index into conditional contents predicated on feature sets `foo` and `bar` and `b.wasm` with an unconditional index into conditional contents predicated on feature sets `foo` and `baz`. Then concatenating `b.wasm` onto `a.wasm` would make `b.wasm`'s unconditional index different for features sets `foo` and `baz`.

I don't think the loss of this property is that important, since in practice we (the tools team) were not seeing much use of the `wasm-merge` tool. In fact, we were seeing so little use that we deleted it from the Binaryen repo and now recommend that users use wasm object files instead.

rossberg commented:
> Without that mechanism we would have to take the first applicable conditional chunk for each section type and we would not have a way to include multiple conditional chunks for a given section.

Wait, why? Why not make them behave like conventional conditional compilation, independent from one another, so multiple conditionals can apply in a single section.

> Consider a.wasm with an unconditional index into conditional contents predicated on feature sets foo and bar and b.wasm with an unconditional index into conditional contents predicated on feature sets foo and baz. Then concatenating b.wasm onto a.wasm would make b.wasm's unconditional index different for features sets foo and baz.

So, if I understand correctly, your scenario in pseudo code is:

a.wasm: (func $0 (call $1)) (#if(foo) (func $1)) (#if(bar) (func $1)) 
b.wasm: (func $0 (call $1)) (#if(foo) (func $1)) (#if(baz) (func $1))

Why would the merge not be something equivalent to the following after proper renumbering:

a+b.wasm: (func $0 (call $2)) (func $1 (call $3)) (#if(foo) (func $2)) (#if(bar) (func $2)) (#if(foo) (func $3)) (#if(bar) (func $3))

That still seems correct, or am I missing something?

> in practice we (the tools team) were not seeing much use of the wasm-merge tool. In fact, we were seeing so little use that we deleted it from the Binaryen repo and now recommend that users use wasm object files instead.

FWIW, I don't think that is necessarily the preferred option for uses outside the Binaryen/LLVM tooling universe. For example, on our platform, where we have an independent compiler, the overhead and complexity of object files would serve no purpose, and we do plain module merging for linking (using our own tool).

tlively commented:
> > Without that mechanism we would have to take the first applicable conditional chunk for each section type and we would not have a way to include multiple conditional chunks for a given section.
> 
> Wait, why? Why not make them behave like conventional conditional compilation, independent from one another, so multiple conditionals can apply in a single section.

Never mind, I solved this problem by defining an active feature set in the new scheme as well. Should work as expected now.

> > Consider a.wasm with an unconditional index into conditional contents predicated on feature sets foo and bar and b.wasm with an unconditional index into conditional contents predicated on feature sets foo and baz. Then concatenating b.wasm onto a.wasm would make b.wasm's unconditional index different for features sets foo and baz.
> 
> So, if I understand correctly, your scenario in pseudo code is:
> 
> a.wasm: (func $0 (call $1)) (#if(foo) (func $1)) (#if(bar) (func $1))
> b.wasm: (func $0 (call $1)) (#if(foo) (func $1)) (#if(baz) (func $1))
> 
> Why would the merge not be something equivalent to the following after proper renumbering:
> 
> a+b.wasm: (func $0 (call $2)) (func $1 (call $3))
> (#if(foo) (func $2)) (#if(bar) (func $2))
> (#if(foo) (func $3)) (#if(baz) (func $3))
> 
> That still seems correct, or am I missing something?

The issue is that if `baz` is the active feature set then there is no `(func $2)`, so the real index of the `baz`-enabled `(func $3)` in that case is 2, not 3.

> > in practice we (the tools team) were not seeing much use of the wasm-merge tool. In fact, we were seeing so little use that we deleted it from the Binaryen repo and now recommend that users use wasm object files instead.
> 
> FWIW, I don't think that is necessarily the preferred option for uses outside the Binaryen/LLVM tooling universe. For example, on our platform, where we have an independent compiler, the overhead and complexity of object files would serve no purpose, and we do plain module merging for linking (using our own tool).

Fair enough. Module merging would still be possible, of course, but it may require adding nop dummy items in some cases, such as the one illustrated above.

rossberg commented:
> The issue is that if baz is the active feature set then there is no (func $2), so the real index of the baz-enabled (func $3) in that case is 2, not 3.

But then a.wasm alone would already have been broken, right? The implicit correctness assumption of a.wasm is that either foo or bar is in the feature set. Likewise either foo or baz for b.wasm. Given these pre-conditions, a+b.wasm would still be correct as well under the combined pre-condition.

Or put differently, there is no way any transformation could make the linked module work if these conditions are not met.

rossberg commented on proposals/feature-detection/Overview.md:
```diff
@@ -2,8 +2,160 @@
+contents replace the unconditional section contents (TODO: what if there are
+multiple active conditional subsections?):
+
+ - Start section
```
This might be a good opportunity to introduce multiple start functions (via multiple start sections). It already is a bummer that they are not allowed (e.g., preventing some merges), but here that restriction really turns into a discontinuity.

tlively replied:
Would the multiple start functions be called sequentially at startup? That seems like a reasonable extension to me for the case where we are merging modules.

binji replied:
Agreed, it would have been nicer to have this match the current section syntax (`vector(funcidx)`), but adding multiple sections is fine too. And yeah, I think it makes sense that they'd be called sequentially.

tlively replied:
Actually, we can update this to be `vector(funcidx)` in a backwards-compatible manner for the same reason we can shove new subsections into all the existing sections: the section size is able to disambiguate.

So the start section contents would become
```
funcidx | vector(funcidx) conditional(vector(funcidx))*
```

rossberg replied:
See above, let's not go there.

rossberg commented on proposals/feature-detection/Overview.md:
```diff
@@ -2,8 +2,160 @@
+multiple active conditional subsections?):
+
+ - Start section
+ - Data count section
```
Maybe it's more useful to make a special rule for the data count section that just sums up the counts?

tlively replied:
So does the data count section become this?
```
count | vector(count) conditional(vector(count))*
```
And then its value is the sum of the active counts?

tlively commented:
> The implicit correctness assumption of a.wasm is that either foo or bar is in the feature set. Likewise either foo or baz for b.wasm. Given these pre-conditions, a+b.wasm would still be correct as well under the combined pre-condition.
> 
> Or put differently, there is no way any transformation could make the linked module work if these conditions are not met.

Yeah that's true. Basically calling any exported functionality from `a` when none of its feature sets are available would necessarily yield unexpected results. A merge tool would probably want to discard any feature sets not in the intersection of supported feature sets of the input modules.

binji commented on proposals/feature-detection/Overview.md:
```diff
@@ -61,82 +61,101 @@ Additional principles motivate the design of this proposal:
-first feature set for which all features are supported by the host engine is the
-*active feature set index*. If no such index exists, the *active feature set
-index* is instead *N*.
+New *conditional subsections* are introduced in a number of existing section
```
I think this needs an example to make it clearer. AIUI, you're planning to use the "unused" part of a section for the conditional subsection, right? In particular, I want to make it clear that a tool that wants to skip either the entire section, or the a feature set can do so.

rossberg replied:
Oh, is that the intention? If so, please reconsider. It is a very intentional property of the current binary format (and one we had heated discussions about) that you never need to look at size fields for disambiguation. That e.g. allows the binary format to be described by a grammar and avoids conflating abstraction levels.

Why can't we simply introduce a new section type of conditional section that contains one regular section?

tlively replied:
Thanks, I didn't know the history there. I think that makes sense, and I think your idea for a new section type containing a normal section makes sense as well. Should each kind of new section type get a new section ID, or should there be only a single new section ID and the contents are interpreted based on the section ID included in the child section? I lean toward the latter to avoid an explosion of IDs.

rossberg replied:
Yes, I think a single section is enough.

In fact, it may be worth considering factorising this even further, into two independent features:

1. Allow all sections to appear multiple times. Define what that means: concatenation for vector-like sections (treating start as a singleton vector), addition for count-like sections.

2. Introduce a conditional section that can be wrapped around any section (anywhere).

This simplifies the semantics by making it less monolithic. It also has a number of additional benefits:

* We don't need to specify anything extra when introducing new section types later (like exceptions).
* The producer gets full control over where in an index space they can place conditional definitions, which may be relevant for some types (e.g., with multiple memories, memory 0 will likely be more efficient).
* It avoids any specific reasoning as to whether conditionals should be allowed for certain sections like imports -- they do what they do independent of section type, and it's the producer's responsibility to use them in a meaningful manner (as @binji pointed out, there could be use cases even for imports/exports).
* It also makes merging modules even easier.

WDYT?

tlively replied:
I like this idea. The separation of concerns and future-compatibility is attractive.

rossberg commented:
I just noticed that there seems to be something fundamental missing from this proposal: AFAICS, it has no way of expressing what to do in the *absence* of a feature. For example, as currently written, I don't think it can express something even as basic as `#if simd ... #else ... #end`. You'll need some form of negation for that.

In fact, when looking at how conditional compilation is often used in C, it's apparent that you'll probably need the basic Boolean connectives: conjunction, disjunction, and negation. Currently, there is only conjunction, while emulating disjunction would require code duplication.

So instead of feature *sets* I suspect that you need Boolean *feature expressions* to be sufficiently complete, e.g. with the following abstract syntax:
```
<cond> ::= <name> | (not <cond>) | (and <cond> <cond>) | (or <cond> <cond>) | true | false
```
where `<name>` essentially acts like `defined(name)` in CPP.

(You could also add else clauses to conditional sections, but in the presence of negation that is redundant, so I'd leave it out for the sake of simplicity.)

WDYT?

tlively commented:
> AFAICS, it has no way of expressing what to do in the _absence_ of a feature. For example, as currently written, I don't think it can express something even as basic as `#if simd ... #else ... #end`. You'll need some form of negation for that.

I would expect this case would be handled by having two feature sets, `{simd128}`, and `{}` (i.e. MVP). In general the last feature set in a module should be the baseline feature set that must be supported in order to use the module, whether that is MVP or something else.

So yes, these feature sets are basically in disjunctive normal form.

Since the conditional sections will be emitted from tools and not necessarily correspond one-to-one with human-written source code, I don't think it is necessary or desirable to allow arbitrary Boolean expressions. Such expressions would at the very least complicate tools' reasoning about feature sets.

rossberg commented:
>> AFAICS, it has no way of expressing what to do in the absence of a feature. For example, as currently written, I don't think it can express something even as basic as #if simd ... #else ... #end. You'll need some form of negation for that.
>
> I would expect this case would be handled by having two feature sets, `{simd128}`, and `{}` (i.e. MVP). In general the last feature set in a module should be the baseline feature set that must be supported in order to use the module, whether that is MVP or something else.

Not sure I follow. Are you saying that the semantics of `{}` actually is the negation of all features *not* listed? I don't see how that would compose, because then there is no efficient and forwards-compatible way to express a feature set that is agnostic of unrelated features.

To give an example, imagine an engine knows features A and B. Some Wasm module is only aware of A and uses feature sets {A} and {} as you suggest. From what you say it can only follow that their meaning would have to be "A and not B" and "not A and not B", respectively. But that's surely not what the module writer intends, they wanted to express "A" and "not A", independent of any B that they didn't even know about.

So there has to be a difference between a feature *not occurring* in a feature set and the *negation* of a feature. Consequently, you need some way to express negation. Or in other words, a feature set needs to be able to express 3 possible dependencies on a feature: present, absent, don't care. And the latter must be the default for features not mentioned, otherwise adding a new feature to an engine would change the meaning of existing modules.

tlively commented:
As written, feature sets are ordered and the first feature set to be a subset of the host's features is chosen to be the active feature set. Each feature set in this scheme could be translated into a logical formula by making the set into a conjunction of (1) its features and (2) the negations of the logical formulae for all previous feature sets. So if the ordered feature sets were `{foo, bar}, {foo}, {}`, these would translate into the following formulae:

1. `foo /\ bar`
1. `foo /\ ~(foo /\ bar)`
1. `true /\ ~foo /\ ~(foo /\ bar)`

rossberg commented:
Ah, I see, thanks. That probably could use some clarification in the PR. But isn’t it even more than that? I think you also implicitly add the negation of all conditions of previous subsections within the same section? Otherwise you couldn't encode an `else`.

Is there any benefit in introducing these implicit negation dependencies everywhere? And doesn’t it require somewhat unnatural ways of expressing things sometimes? I have to say that it is a bit confusing, at least to me.

For example, if I had code whose activating condition was `(foo /\ bar) \/ (foo /\ baz)`, or in pseudo code,
```
#if((foo /\ bar) \/ (foo /\ baz)) … #else … #endif
```
then IIUC, I’d have to express that with the feature sets `{foo, bar, baz}, {foo, bar}, {foo, baz}` for the first condition and the empty list for the else? It took me a moment to figure that out. The reason for the repetition night not be immediately obvious to the reader either.

Another thing is that this interpretation wouldn’t work anymore with the factorisation I suggested above, because the dependent meaning between the first and the second feature set list could no longer be injected when all conditional sections are independent.

I think all that could be solved and simplified by pushing through the idea of disjunctive normal form to its logical conclusion. All you’d need to add is a Boolean flag next to each feature name that says whether it is negated. That is, a `feature_set` is
```
features  |  feature*
```
And a `feature` is
```
negated | u8 
name     | name
```
So, condition = disjunction of feature sets; feature set = conjunction of features; feature = possibly negated name.

With that, every feature set and every condition can be given an independent and self-contained meaning. The condition above could be expressed as `{foo, bar}, {foo baz}`, like I would expect. I would argue that is easier to understand and easier to get right.

Does that make sense?

tlively commented:
Yes, what you propose makes sense, but I worry that it is more complicated than we need or want. When you consider modeling traditional conditional compilation with `#ifdefs` and whatnot, I agree that using logical formulae in some normal form is the simplest solution, but I don't believe that is what producers and consumers of this feature will aim to model.

Conditional compilation in the style of `#ifdefs` happens in the frontend of the toolchain, so it would be surprising to users (and hard to implement in the tools) if code in both arms of an `#ifdef`-like mechanism persisted into the final binaries. The normal way to use such mechanisms is to have multiple builds producing different binaries, and although that may be old-fashioned and suboptimal, I don't think we should impose a change in that model on existing toolchains.

Instead, the user-facing feature I would most like to enable with conditional sections is more similar to ELF's ifunc mechanism. The user would write something like the following, which can already be written using GCC/clang attributes:

```C
__attribute__((target("simd128,nontrapping-fptoint")))
int foo(void) { ... }

__attribute__((target("nontrapping-fptoint")))
int foo(void) { ... }

__attribute__((target("default")))
int foo(void) { ... }
```

Notice that the user interface corresponds almost directly to the feature set scheme. That means there is also precedent for compilers, linkers, and loaders to reason about feature sets in this way. Not to say we have to do things this way because that is how ELF does it, but there would be a semantic gap to overcome in the tools and for users trying to reason about their modules at a low level if we did decide to go with a different model.

---

Addressing your concerns with the feature set mechanism:


> ```
> #if(foo \/ ~bar) A #else B #endif
> ```
> 
> then IIUC, I’d have to express that with _three_ conditions, with respective feature set lists `{bar}`, `{foo}`, and `{}`, or in pseudo code,
> 
> ```
> #if(bar) B #elseif(foo) A #else B #endif
> ```
> 
> Note the code duplication of `B`. At least I can't think of a better way.

No code duplication is requires when conditional sections can be activated by multiple feature sets. In this example, B would be activated by either `{bar}` or `{}` and `A` would be activated by `{foo}`. However, this depends on `{foo}` having a higher precedence than `{}`. Establishing this precedence was one of the roles of the separate feature set section present in my original writeup.

> Another thing is that this interpretation wouldn’t work anymore with the factorisation I suggested above, because the dependent meaning between the first and the second feature set list could no longer be injected when all conditional sections are independent. You'd need at least an else clause.

This is also solved by the separate target features section, and was the intuitive basis for my originally including it. What do you think of keeping the feature set mechanism but reintroducing the feature set section to solve the problems you've identified? That is of course an extra complication, but I think that on the whole it still has less of a semantic gap with how we would like users and tools to model conditional sections than more general logical formulae.

rossberg commented:
> Yes, what you propose makes sense, but I worry that it is more complicated than we need or want. 

Ah, but I am actually trying to suggest an overall simplification! It's adding the Boolean flag, but that allows structurally simplifying conditonal sections themselves to a single condition with code as opposed to a dependent vector of pairs of them. Effectively, it is trading built-in elif chains for a simple negation operator.

More importantly, that is already expressively complete and doesn't require the further complication of the feature section to solve the code duplication problem (which I think needs to be solved).

> there would be a semantic gap to overcome in the tools and for users trying to reason about their modules at a low level if we did decide to go with a different model.

You have a much more complete picture of the tool chain than I, but I wonder what would prevent mapping this model to attributes directly, e.g., allowing
```C
__attribute__((target("~simd128,nontrapping-fptoint")))
int foo(void) { ... }
```
Assuming this works, the structural simplification I mention above would map to attributes more directly than elif chains would, I think.

In contrast to this, I don't yet understand how the indirection through an extra feature section would be expressed in attributes.

tlively commented:
Ok, let's go with your suggestion, at least until we can evaluate the additional complexity from implementing it that way in the tools. I agree that your suggestions simplify the specification and engine implementations, but ultimately most of the implementation work for this proposal will lie in the tools. Once we have an initial prototype we should make final design decisions with the goal of minimizing total complexity, and the tools implementation should weight heavily in that judgement.

rossberg commented:
Cool, thanks! Makes sense. My assumption would be that tools on the producer side still have a lot of leeway how to design a feature test mechanism -- ultimately, the difference between the representations we've been discussing is merely a simple "lowering" step into plain conditionals, based on the [interpretation you gave](https://github.com/WebAssembly/feature-detection/pull/2#issuecomment-518793161), which a producer could apply during codegen.

tlively commented:
@rossberg PTAL. I'm still not sure it's ideal to have the linker be normalizing Boolean formulae, but I'm happy to merge this initial design into the repo and solicit feedback from the broader community. Perhaps a quick update for the CG would be a reasonable next step.

tlively commented:
@rossberg ping! Do you have any more small nits here? Further high-level discussion can be done on follow-up issues.

rossberg commented on proposals/feature-detection/Overview.md:
```diff
@@ -61,85 +53,137 @@ Additional principles motivate the design of this proposal:
+There are two parts to this design. First, the binary format is extended to
+allow many section types to appear multiple times. Second, a new type of
+conditional section is introduced that can wrap any other section except
+itself. During validation, the contents of conditional sections are considered
```
Nit: it's not just during validation. Conceptually, I'd phrase this somehow to say that conditional sections are resolved during decoding already.

rossberg commented on proposals/feature-detection/Overview.md:
```diff
@@ -61,85 +53,137 @@ Additional principles motivate the design of this proposal:
+
+New *conditional sections* are introduced to wrap other sections. Conditional
+sections contain a *predicate* and binary *contents*. The contents of a
+conditional section are skipped entirely during parsing and validation unless
```
Nit: See above, I think they won't even be part of the abstract syntax, but dropped during decoding already. (Secondary nit: the spec speaks of decoding for the binary format, parsing for the text format.)

rossberg commented on proposals/feature-detection/Overview.md:
```diff
@@ -61,85 +53,137 @@ Additional principles motivate the design of this proposal:
+conditional section are skipped entirely during parsing and validation unless
+the features supplied by the host satisfy the conditional section's
+predicate. If the predicate is satisfied, however, the binary contents are
+parsed and validated as a section. It is a validation error if a conditional
```
Nit: This would be a decoding error. To avoid pedantry, maybe just say that a conditional section is malformed if it contains another conditional section.

rossberg replied:
Ah, the binary grammar simply doesn't allow conditional sections nested into conditional sections, no further interpretation necessary.

(The reason I am nitpicking on this is that sections in general are considered an artefact of the binary format only and don't appear in the AST (or text format). So validation, which operates on the AST (conceptually), doesn't even have a way to talk about them. Accordingly, conditional sections are a form of decoding-time preprocessing unknown to the rest of the semantic pipeline. Does that make sense?)

rossberg commented on proposals/feature-detection/Overview.md:
```diff
@@ -61,85 +53,137 @@ Additional principles motivate the design of this proposal:
+component `feature_set`s are satisfied. A `feature_set` is satisfied if all of
+its component `feature`s are satisfied. A `feature` is satisfied if its
+`negated` field is 0 and the host supplies the feature or if its `negated` field
+is 1 and the host does not supply the feature. It is a validation error if a
```
Nit: Also malformed, not a validation error.
